x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/DataFrameAnalyticsSource.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ml.dataframe;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.xpack.core.ml.job.messages.Messages;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.ml.utils.QueryProvider;
+import org.elasticsearch.xpack.core.ml.utils.XContentObjectTransformer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class DataFrameAnalyticsSource implements Writeable, ToXContentObject {
+
+    public static final ParseField INDEX = new ParseField("index");
+    public static final ParseField QUERY = new ParseField("query");
+
+    public static final ConstructingObjectParser<DataFrameAnalyticsSource, Void> STRICT_PARSER = createParser(false);
+    public static final ConstructingObjectParser<DataFrameAnalyticsSource, Void> LENIENT_PARSER = createParser(true);
+
+    public static ConstructingObjectParser<DataFrameAnalyticsSource, Void> createParser(boolean ignoreUnknownFields) {
+        ConstructingObjectParser<DataFrameAnalyticsSource, Void> parser = new ConstructingObjectParser<>("data_frame_analytics_source",
+            ignoreUnknownFields, a -> new DataFrameAnalyticsSource((String) a[0], (QueryProvider) a[1]));
+        parser.declareString(ConstructingObjectParser.constructorArg(), INDEX);
+        parser.declareObject(ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> QueryProvider.fromXContent(p, ignoreUnknownFields, Messages.DATA_FRAME_ANALYTICS_BAD_QUERY_FORMAT), QUERY);
+        return parser;
+    }
+
+    private final String index;
```

Comment:
transforms supports lists: `String[] index` - would be good to be consistent? Or is there a good reason to only support 1 index?, it seems multi indices using wildcards is supported

Reply:
I haven't thought through if this can work with multiple indices as source. I think it can. I'll raise a ticket for it and make the changes in a different PR.